File: py/src/piper_py/runner.py
from __future__ import annotations

import os
import subprocess
import sys
from pathlib import Path


def clean_env() -> dict[str, str]:
    env = os.environ.copy()
    env.pop("VIRTUAL_ENV", None)
    tool_bin = str(Path(sys.prefix) / "bin")
    env["PATH"] = tool_bin + os.pathsep + env.get("PATH", "")
    return env


def run_command(name: str, command: str) -> tuple[bool, str]:
    completed = subprocess.run(
        command, shell=True, capture_output=True, text=True, env=clean_env()
    )
    if completed.returncode == 0:
        return True, f"OK   {name}"
    output = (completed.stdout + completed.stderr).rstrip()
    return False, f"FAIL {name}\nCOMMAND {command}\n{output}"


def run_commands(commands: list[tuple[str, str]]) -> tuple[int, list[str]]:
    any_failed = False
    outputs: list[str] = []
    for name, command in commands:
        passed, output = run_command(name, command)
        outputs.append(output)
        if not passed:
            any_failed = True
    return 2 if any_failed else 0, outputs

File: ts/src/runner.ts
import { execSync } from "node:child_process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const packageRoot = dirname(__dirname);
const nodeModulesBin = join(packageRoot, "node_modules", ".bin");

export function cleanEnv(): Record<string, string> {
  const env = { ...process.env } as Record<string, string>;
  env.PATH = nodeModulesBin + ":" + (env.PATH ?? "");
  return env;
}

export function runCommand(name: string, command: string): [boolean, string] {
  try {
    execSync(command, { stdio: "pipe", env: cleanEnv() });
    return [true, `OK   ${name}`];
  } catch (err: unknown) {
    const e = err as { stdout?: Buffer; stderr?: Buffer };
    const stdout = e.stdout?.toString() ?? "";
    const stderr = e.stderr?.toString() ?? "";
    const output = (stdout + stderr).trimEnd();
    return [false, `FAIL ${name}\nCOMMAND ${command}\n${output}`];
  }
}

export function runCommands(
  commands: [string, string][],
): [number, string[]] {
  let anyFailed = false;
  const outputs: string[] = [];

  for (const [name, command] of commands) {
    const [passed, output] = runCommand(name, command);
    outputs.push(output);
    if (!passed) {
      anyFailed = true;
    }
  }

  return [anyFailed ? 2 : 0, outputs];
}
